sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramEndpoints.scala

```diff
@@ -0,0 +1,465 @@
+    // invalid, size of bins
+    var len: Int = Ints.BYTES + Ints.BYTES
+    obj.bins.foreach { case (key, value) =>
+      // length of key, key, value
+      len += Ints.BYTES + key.getBytes.length + Longs.BYTES
+    }
+    len
+  }
+
+  final def serialize(obj: StringEndpointsDigest): Array[Byte] = {
+    val buffer = ByteBuffer.wrap(new Array(length(obj)))
+    buffer.putInt(if (obj.invalid) 0 else 1)
+    buffer.putInt(obj.bins.size)
+    obj.bins.foreach { case (key, value) =>
+      val bytes = key.getBytes
+      buffer.putInt(bytes.length)
+      buffer.put(bytes)
+      buffer.putLong(value)
+    }
+    buffer.array()
+  }
+
+  final def deserialize(bytes: Array[Byte]): StringEndpointsDigest = {
+    val buffer = ByteBuffer.wrap(bytes)
+    val invalid = if (buffer.getInt == 0) true else false
+    val size = buffer.getInt
+    val bins = new mutable.HashMap[UTF8String, Long]
+    var i = 0
+    while (i < size) {
+      val keyLength = buffer.getInt
+      var j = 0
+      val keyBytes = new Array[Byte](keyLength)
+      while (j < keyLength) {
+        keyBytes(j) = buffer.get()
+        j += 1
+      }
+      val value = buffer.getLong
+      bins.put(UTF8String.fromBytes(keyBytes), value)
+      i += 1
+    }
+    new StringEndpointsDigest(bins, invalid)
+  }
+}
+
+/**
+ * Digest class for column of numeric type.
+ * @param bins A HashMap to maintain frequency of each distinct value.
+ * @param percentileDigest A helper class to compute approximate percentiles.
+ */
+case class NumericEndpointsDigest(
+    percentileDigest: PercentileDigest,
+    bins: mutable.HashMap[Double, Long] = mutable.HashMap.empty[Double, Long])
+  extends EndpointsDigest {
+
+  var mapInvalid = false
+
+  def this(
+      percentileDigest: PercentileDigest,
+      bins: mutable.HashMap[Double, Long],
+      invalid: Boolean,
+      mapInvalid: Boolean) = {
+    this(percentileDigest, bins)
+    this.invalid = invalid
+    this.mapInvalid = mapInvalid
+  }
+
+  override def update(dataType: DataType, value: Any, numBins: Int): Unit = {
+    if (dataType.isInstanceOf[DecimalType]) {
+      val decimal = value.asInstanceOf[Decimal]
+      val double = decimal.toDouble
+      if (double == Double.PositiveInfinity || double == Double.NegativeInfinity) {
+        // This value has too great a magnitude to represent as a Double.
+        invalid = true
+        clear()
+        return
+      }
+    }
+
+    // We use Double to represent endpoints (in histograms) for simplicity.
+    // Loss of precision is acceptable because we are computing approximate answers anyway.
+    val doubleValue = dataType match {
+      case n: NumericType =>
+        n.numeric.toDouble(value.asInstanceOf[n.InternalType])
+      case d: DateType =>
+        value.asInstanceOf[Int].toDouble
+      case t: TimestampType =>
+        value.asInstanceOf[Long].toDouble
+    }
+    // update percentileDigest
+    percentileDigest.add(doubleValue)
```

Comment:
I take it you're updating the percentile in all cases because you don;t want to merge maps with percentiledigests ?